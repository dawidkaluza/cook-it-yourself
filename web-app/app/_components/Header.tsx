import React from "react";
import {isSignedIn} from "@/app/_api/auth";
import Link from "next/link";

const Header = () => {
  const signedIn = isSignedIn();

  return (
    <header className="navbar navbar-expand bg-body-tertiary">
      <div className="container-fluid text-center">
        <Link href="/" className="navbar-brand">Cook it yourself</Link>
        {signedIn
          ? <SignOutButton />
          : <SignInButton />
        }
      </div>
    </header>
  );
}

const SignInButton = () => {
  // noinspection HtmlUnknownTarget
  return (
    <a
      href="/sign-in"
      className="btn btn-primary btn-sm"
    >
      Sign in
    </a>
  );
};


const SignOutButton = () => {
  // noinspection HtmlUnknownTarget
  return (
    <a
      href="/sign-out"
      className="btn btn-primary btn-sm"
    >
      Sign out
    </a>
  );
};

export {Header};