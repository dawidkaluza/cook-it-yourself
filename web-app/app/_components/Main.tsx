import React from "react";

const Main = ({
  children
}: Readonly<{
  children: React.ReactNode
}>) => {
  return (
    <main style={{margin: "1em"}}>
      {children}
    </main>
  );
};

export { Main };