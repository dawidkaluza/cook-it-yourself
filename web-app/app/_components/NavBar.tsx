"use client"

import React from "react";
import Link from "next/link";
import {usePathname} from "next/navigation";
import {useAuth} from "@/app/_hooks/auth/useAuth";

const NavBar = () => {
  const { isSignedIn } = useAuth();

  return (
    <nav className="navbar navbar-expand bg-body-tertiary">
      <div className="container-fluid text-center">
        <Link href="/" className="navbar-brand">Cook it yourself</Link>
        {isSignedIn ? <NavBarMenu /> : <SignInButton />}
      </div>
    </nav>
  );
};

const NavBarMenu = () => {
  return (
    <>
      <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar-menu"
              aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
        <span className="navbar-toggler-icon"></span>
      </button>

      <div className="collapse navbar-collapse" id="navbar-menu">
        <ul className="navbar-nav">
          <NavBarItem path="/my-recipes">My recipes</NavBarItem>
          <NavBarItem path="/cooking">Cooking</NavBarItem>
        </ul>
      </div>
    </>
  );
};

const NavBarItem = ({
  path,
  children
}: Readonly<{
  path: string,
  children: React.ReactNode,
}>) => {
  const currentPath = usePathname();

  return (
    <li className="nav-item">
      <Link className={path === currentPath ? "nav-link active" : "nav-link"} href={path}>{children}</Link>
    </li>
  );
};

const SignInButton = () => {
  return (
    <a
      href={process.env.API_GATEWAY_CLIENT_URL + "/oauth2/authorization/ciy"}
      className="btn btn-primary btn-sm"
    >
      Sign in
    </a>
  );
};

export {NavBar};