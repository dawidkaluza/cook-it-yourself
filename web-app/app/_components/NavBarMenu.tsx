"use client";

import React from "react";
import {usePathname} from "next/navigation";
import Link from "next/link";

export const NavBarMenu = () => {
  return (
    <>
      <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar-menu"
              aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
        <span className="navbar-toggler-icon"></span>
      </button>

      <div className="collapse navbar-collapse" id="navbar-menu">
        <ul className="navbar-nav">
          <NavBarItem path="/my-recipes">My recipes</NavBarItem>
        </ul>
      </div>
    </>
  );
};

const NavBarItem = ({
  path,
  children
}: Readonly<{
  path: string,
  children: React.ReactNode,
}>) => {
  const currentPath = usePathname();

  return (
    <li className="nav-item">
      <Link className={path === currentPath ? "nav-link active" : "nav-link"} href={path}>{children}</Link>
    </li>
  );
};