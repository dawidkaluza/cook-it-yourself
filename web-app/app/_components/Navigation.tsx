"use client";

import {usePathname} from "next/navigation";
import Link from "next/link";

const Navigation = () => {
  const isNumeric = (num: any) => (typeof(num) === 'number' || typeof(num) === "string" && num.trim() !== '') && !isNaN(num as number);
  const toPathName = (segment: string) => {
    if (isNumeric(segment)) {
      return "View";
    }

    const name = segment.replace("-", " ").trim();
    return name.charAt(0).toUpperCase() + name.slice(1);
  }

  const pathname = usePathname();

  const navItems = new Array<{ name: string; path: string, active: boolean }>();
  const isHomeActive = pathname.length === 1;
  navItems.push({ name: "Home", path: "/", active: isHomeActive });

  if (!isHomeActive) {
    let lastIndexOfSeparator = 0;
    let indexOfSeparator = 0;
    do {
      indexOfSeparator = pathname.indexOf("/", lastIndexOfSeparator + 1);
      const segmentLastPos = indexOfSeparator !== -1 ? indexOfSeparator : pathname.length;
      const segment = pathname.slice(lastIndexOfSeparator + 1, segmentLastPos);
      navItems.push({ name: toPathName(segment), path: pathname.slice(0, segmentLastPos), active: indexOfSeparator === -1 });
      lastIndexOfSeparator = indexOfSeparator;
    } while (indexOfSeparator !== -1)
  }

  return (
    <nav className="m-3">
      <ol className="breadcrumb">
        {navItems.map(navItem => {
          return navItem.active ? (
            <li key={navItem.path} className="breadcrumb-item active">
              {navItem.name}
            </li>
          ) : (
            <li key={navItem.path} className="breadcrumb-item">
              <Link href={navItem.path}>{navItem.name}</Link>
            </li>
          );
        })}
      </ol>
    </nav>
  );
};

export {Navigation};