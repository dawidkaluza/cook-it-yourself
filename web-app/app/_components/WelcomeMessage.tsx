"use client";

import {useAuth} from "@/app/_api/hooks";

const WelcomeMessage = () => {
  const {isSignedIn, name} = useAuth();

  if (isSignedIn) {
    return (
      <>
        <p className="text-center">
          Welcome, {name}!
        </p>
        <p className="text-center">
          Use the navigation above to start your cooking journey.
        </p>
      </>
    );
  }

  return (
    <>
      <p className="text-center">
        Manage recipes, rate dishes, plan diet, plan grocery shopping and more.
      </p>
      <p className="text-center">
        Want to try it out? <a href="/sign-in">Sign in</a>.
      </p>
    </>
  );
};

export {WelcomeMessage}