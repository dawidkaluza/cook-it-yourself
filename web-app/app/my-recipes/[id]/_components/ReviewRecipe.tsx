import {reviewRecipe} from "@/app/my-recipes/[id]/actions";
import Link from "next/link";

type Props = {
  id: number;
};

const ReviewRecipe = async ({ id } : Props) => {
  const recipe = await reviewRecipe(id);

  return (
    <div className="container-fluid">
      <div className="row">
        <div className="col-md-8 offset-md-2">
          <div>
            <h1>{recipe.name}</h1>
            <h6>{recipe.description}</h6>
            <ul className="list-unstyled">
              <li>Portion size: {Number(recipe.portionSize.value) + " " + recipe.portionSize.measure}</li>
              <li>Cooking time: {recipe.cookingTime} minutes</li>
            </ul>
          </div>

          <div className="mt-4">
            <h2>Ingredients</h2>
            <ul>
              {recipe.ingredients.map((ingredient) => (
                <li key={ingredient.id}>
                  {ingredient.name + " " + Number(ingredient.value) + " " + ingredient.measure}
                </li>
              ))}
            </ul>
          </div>

          <div className="mt-4">
            <h2>Method steps</h2>

            {recipe.methodSteps.map(step => (
              <p key={step.id}>{step.text}</p>
            ))}
          </div>
        </div>
      </div>
      <div className="row">
        <div className="col-md-8 offset-md-2">
          <hr/>
        </div>

        {/* TODO test this part */}
        <div className="col-md-8 offset-md-2 d-flex justify-content-center justify-content-sm-start">
          <Link href={`/my-recipes/${recipe.id}/edit`} className="btn btn-primary mx-2">Modify</Link>
          <Link href={`/my-recipes/${recipe.id}/delete`} className="btn btn-danger mx-2">Delete</Link>
        </div>
      </div>
    </div>
  );
};

export {ReviewRecipe};