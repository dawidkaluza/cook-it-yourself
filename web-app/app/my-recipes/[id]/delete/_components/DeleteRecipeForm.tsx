"use client";

import {deleteRecipe} from "@/app/my-recipes/[id]/delete/actions";
import Link from "next/link";
import {useFormState} from "react-dom";

const DeleteRecipeForm = ({ id } : { id: number }) => {
  let [error, action] = useFormState(deleteRecipe, "");

  return (
    <form action={action} noValidate>
      <input type="hidden" name="id" value={id} />
      <p>
        Are you sure that you want to delete the recipe?
      </p>

      <div className="my-4 d-flex justify-content-center btn-group">
        <div className="btn-group">
          <button className="btn btn-danger" type="submit">
            <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" className="bi bi-x"
                 viewBox="0 0 16 16">
              <path
                d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"/>
            </svg>

            {" Delete "}
          </button>
          <Link href={`/my-recipes/${id}`} className="btn btn-outline-secondary">
            Cancel
          </Link>
        </div>
      </div>

      {error && (
        <p className="text-danger">
          {error}
        </p>
      )}
    </form>
  );
}

export {DeleteRecipeForm};