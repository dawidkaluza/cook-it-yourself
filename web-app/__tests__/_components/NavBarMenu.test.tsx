import {render, screen} from "@testing-library/react";
import {NavBarMenu} from "@/app/_components/NavBarMenu";

describe("NavBarMenu component", () => {
  test("renders", () => {
    // Given, when
    const menu = render(<NavBarMenu />);

    // Then
    const button = screen.getByRole("button");
    expect(button).not.toBeNull();
    expect(button?.getAttribute("data-bs-toggle")).toBe("collapse");
    expect(button?.getAttribute("data-bs-target")).toBe("#navbar-menu");

    const menuContainer = menu.container.querySelector("#navbar-menu");
    expect(menuContainer).not.toBeNull();
    expect(menuContainer?.getAttribute("class")).toBe("collapse navbar-collapse");

    const links = menu.queryAllByRole("link");
    expect(links.length).toBe(2);
    expect(links[0]).not.toBeNull();
    expect(links[0].getAttribute("href")).toBe("/my-recipes");
    expect(links[1]).not.toBeNull();
    expect(links[1].getAttribute("href")).toBe("/cooking");

    expect(menu.container).toMatchSnapshot();
  });
})