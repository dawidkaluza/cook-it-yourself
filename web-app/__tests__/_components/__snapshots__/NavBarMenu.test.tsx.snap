// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavBarMenu component > renders 1`] = `
<div>
  <button
    aria-controls="navbarNav"
    aria-expanded="false"
    aria-label="Toggle navigation"
    class="navbar-toggler"
    data-bs-target="#navbar-menu"
    data-bs-toggle="collapse"
    type="button"
  >
    <span
      class="navbar-toggler-icon"
    />
  </button>
  <div
    class="collapse navbar-collapse"
    id="navbar-menu"
  >
    <ul
      class="navbar-nav"
    >
      <li
        class="nav-item"
      >
        <a
          class="nav-link"
          href="/my-recipes"
        >
          My recipes
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link"
          href="/cooking"
        >
          Cooking
        </a>
      </li>
    </ul>
  </div>
</div>
`;
